# Review(2)
Thanks for your update. Now the project is accepted! I left one small remark, which does not affect the fact that project is accepted. 


Good luck in future learning!

Hello, my name is Ivan Alexeev and I am going to review your project.


There may be some shortcomings in the work that I will ask you to eliminate, you fix them and I check your decisions. You can find my comments in <font color='green'>green</font>, <font color='orange'>orange</font> or <font color='red'>red</font> boxes like this:


<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: if everything is done successfully
</div>


<div class="alert alert-warning" style="box-shadow: 4px 4px 4px">
Remark: if I can give some recommendations or additional information
</div>


<div class="alert alert-danger" style="box-shadow: 4px 4px 4px">
Need fixing: if the block requires some corrections. Work can't be accepted with the red comments
</div>


Thank you for taking time to complete this project, I appreciate the amount of work you've done! There are some issues that you need to work on, but overall it is a great start! 


Please, don't delete my comments) 


# General feedback
First of all, it is a great start! You've done an amazing job! To begin with, I would like to note that all the calculations were made correctly, the hypotheses are correct, the correct answers were obtained, and simply unrealistic conclusions were written. Few people will write such a good conclusion even in real reports. Continue in the same spirit and try to write a little more information before each step, a small plan.


Unfortunately, there is a few problems in your work. They are not a mistakes. Please, change the design of a notebook. It is inappropriate in real reports to write so many information in one code cell. It is better to separate them. One of the possible ways to do it is separate into parts where only one output with a description. 


Good luck! I\ll wait for your corrections) 

# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

<div class="alert alert-warning" style="box-shadow: 4px 4px 4px">
<s>Remark: in real reports it is better to write some more information in the introduction. For instance, a plan can be written or you can describe a dataset. </s>
</div>

## Open the data file and have a look at the general information. 

***Task: Prepare a report that evaluates if marital status and number of children has an impact on whether a client will default on a loan***

The bank loan's division wants the data to be cleaned up with some instructions:
- Appropriately process artifacts
- Fill in (not delete) missing values
- Delete duplicates
- Perform an analysis and provide conclusions

Questions provided by bank loan's division:
- Is there a connection between having kids and repaying a loan on time?
- Is there a connection between marital status and repaying a loan on time?
- Is there a connection between income level and repaying a loan on time?
- How do different loan purposes affect on-time loan repayment?

***Plan of action***
- Examine the data's general information
- Process obvious artifacts, such as negative numbers, poorly described column names, etc.
    - i.e. evaluate the unique values in appropriate columns for interesting or non-sensical data
- Determine if there are missing values
    - Determine the best/appropriate way and data to fill in missing values
- Determine if there are duplicate values or rows
    - Determine the best/appropriate way to handle duplicates
- Change datatype were appropriate/explained to do so
- Categorize data where appropriate for easier analysis
- Perform analysis to answer questions above

***Import libraries and packages***

In [5]:
# import libraries and packages
import pandas as pd
import nltk
from nltk.stem import WordNetLemmatizer
from collections import Counter
from nltk.stem import SnowballStemmer

***Review data's general information***

In [6]:
# read and save original data
scoring_credit_orig = pd.read_csv('/datasets/credit_scoring_eng.csv')

# review general information
print('First 20 rows of data:')
print(scoring_credit_orig.head(20), '\n')

First 20 rows of data:
    children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary edu

In [7]:
print('Data info:')
scoring_credit_orig.info()

Data info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [8]:
# Determine number of missing rows
print('Number of rows with missing values:')
print(scoring_credit_orig.isnull().sum(), '\n')

Number of rows with missing values:
children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64 



In [9]:
# Determine what columns contain missing values and how much of the data it impacts
missing_per_column = (scoring_credit_orig.isnull().sum()) / (len(scoring_credit_orig))
print('Conversion rate of missing values per column:\n{}'.format(missing_per_column), '\n')

Conversion rate of missing values per column:
children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64 



In [10]:
# Determine the number of duplicate rows
print('Number of duplicate rows: ', scoring_credit_orig.duplicated().sum(), '\n')

Number of duplicate rows:  54 



***Take a closer look at the subcategories in appropriate columns and process artifacts.***

Updates will need to be made to the data frame, but the original needs to be presereved.
Create a new data frame for making changes.

In [11]:
# read and save working version of data
scoring_credit = pd.read_csv('/datasets/credit_scoring_eng.csv')

From looking at the original data, the days_employed column contains artifacts:
- There are a lot of negative values
    - Change these to positive values
- It is meant to represent how long the client has been working, however if these values are meant to represent number of days then some of them are too large.
    - This artifact is not addressed in this investigation.

In [12]:
# Create a new column for days_employed with positive values
scoring_credit['days_employed'] = scoring_credit['days_employed'].abs()

# Confirm column is updated information
print('Updated data:')
print(scoring_credit.head(10))

Updated data:
   children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  Secondary Education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  SECONDARY EDUCATION             1   
8         2    6929.865299         35    BACHELOR'S DEGREE             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     

In [13]:
# Evaluate children subcategories
print('children counts:')
print(scoring_credit['children'].value_counts(), '\n')

children counts:
 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64 



The children column contains artifacts:
- The column represents number of children 
    - Change the column name to better represent this information
- There are negative values
- There is a rather high value

While -1 and 20 may be inappropriate values for number of children, with out being able to ask the investigators clarifying questions about the data and collection/transfer methods I will not change them.
- -1 children may represent clients who have lost children
- 20 children may be a large number of children, but is not impossible

In [14]:
# Rename children column to better represent the column information
scoring_credit.rename(columns={'children': 'num_children'}, inplace=True)

# Confirm column name change
print('Updated data with column name change:')
print(scoring_credit.head())

Updated data with column name change:
   num_children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  Secondary Education             1   
3             3    4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.57

In [15]:
# Evaluate dob_years subcategories
print('dob_years counts:')
print(scoring_credit['dob_years'].value_counts(), '\n')

dob_years counts:
35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64 



In [16]:
# Evaluate education subcategories
print('education counts:')
print(scoring_credit.groupby('education_id')['education'].value_counts(), '\n')

education counts:
education_id  education          
0             bachelor's degree       4718
              BACHELOR'S DEGREE        274
              Bachelor's Degree        268
1             secondary education    13750
              SECONDARY EDUCATION      772
              Secondary Education      711
2             some college             668
              Some College              47
              SOME COLLEGE              29
3             primary education        250
              PRIMARY EDUCATION         17
              Primary Education         15
4             graduate degree            4
              GRADUATE DEGREE            1
              Graduate Degree            1
Name: education, dtype: int64 



The education column contains artifacts:
- Upper and lower case subcategories
- There are multiple subcategories representing the same thing because of case discrepencies

These artifacts will be addressed later in the notebook

In [17]:
# Evaluate family_status subcategories
print('family_status counts:')
print(scoring_credit.groupby('family_status_id')['family_status'].value_counts(), '\n')

family_status counts:
family_status_id  family_status    
0                 married              12380
1                 civil partnership     4177
2                 widow / widower        960
3                 divorced              1195
4                 unmarried             2813
Name: family_status, dtype: int64 



In [18]:
# Evaluate gender subcategories
print('gender counts:')
print(scoring_credit['gender'].value_counts(), '\n')

gender counts:
F      14236
M       7288
XNA        1
Name: gender, dtype: int64 



In [19]:
# Evaluate income_type subcategories
print('income_type counts:')
print(scoring_credit['income_type'].value_counts(), '\n')

income_type counts:
employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
entrepreneur                       2
unemployed                         2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64 



In [20]:
# Evaluate debt subcategories
print('debt counts:')
print(scoring_credit['debt'].value_counts(), '\n')

debt counts:
0    19784
1     1741
Name: debt, dtype: int64 



The information provided in the debt column is meant to represent if a client has defaulted on a loan before.

These values are not very informative

Create a new column to better represent this information
   - update with a name that represents the info
   - change the values to be more descriptive and readable

In [21]:
# Create a more readable column for those that have defaulted on a loan before:
# Create a function to change the column values
def assign_default_status(default_status):
    if default_status == 1:
        return 'Yes'
    if default_status == 0:
        return 'No'
    else:
        return 'Unknown'

# Create new column with newly assigned default history
scoring_credit['defaulted_before'] = scoring_credit['debt'].apply(assign_default_status)
# Check data with new column and updated information
print('Updated data with new column and updated information:')
print(scoring_credit.head(), '\n')

Updated data with new column and updated information:
   num_children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  Secondary Education             1   
3             3    4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree   

In [22]:
# Confirm counts match between the old and new columns
print('defaulted_before subcategory totals:')
print(scoring_credit['defaulted_before'].value_counts(), '\n')

defaulted_before subcategory totals:
No     19784
Yes     1741
Name: defaulted_before, dtype: int64 



In [23]:
# Evaluate the percentage of clients who have defaulted
defaulted_clients = 0
for row in scoring_credit['debt']:
    if row == 1:
        defaulted_clients += 1

perc_defaulted_clients = defaulted_clients / len(scoring_credit)
print('Percentage of defaulted clients: {:.1%}'.format(perc_defaulted_clients), '\n')

Percentage of defaulted clients: 8.1% 



In [24]:
# Evaluate purpose subcategories
print('purpose counts:')
print(scoring_credit['purpose'].value_counts())

purpose counts:
wedding ceremony                            797
having a wedding                            777
to have a wedding                           774
real estate transactions                    676
buy commercial real estate                  664
housing transactions                        653
buying property for renting out             653
transactions with commercial real estate    651
housing                                     647
purchase of the house                       647
purchase of the house for my family         641
construction of own property                635
property                                    634
transactions with my real estate            630
building a real estate                      626
buy real estate                             624
purchase of my own house                    620
building a property                         620
housing renovation                          612
buy residential real estate                 607
buying my own car       

The purpose column contains artifacts:
- Duplicate values in a way
    - There are several purposes provided and many of them represent the same thing but phrased differently

These artifacts will be addressed later in the notebook

***Data frame info with updates***

In [25]:
# Review data info
print('Dataframe info with updates:')
scoring_credit.info()

Dataframe info with updates:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
num_children        21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
defaulted_before    21525 non-null object
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


<div class="alert alert-warning" style="box-shadow: 4px 4px 4px">
<s>Remark: it is not a mistake, however it is accepted to import libraries in a separate code cell. Also it is a little bit better to use each method in a different code cells. It will just look prettier</s>
</div>

### Conclusion

***General info:***
- There are 21525 entries and 12 columns
- The columns represent categorical and numerical types
- Each row represents a specific customer
    - There is not a unique identifier for each customer, so there could be duplicates representing different custormers
- Columns days_employed and total_income contain missing data: 10% of the total entries
- Columns days_employed has a lot of negative numbers, which doesnt make sense, and some values are too high for a human lifespan
    - Perhaps there is a calculation error that happened or this column is mislabeled and is meant to represent different information
- Column education contains lower and upper case entries
- There are 54 duplicates
- Some of the column names are hard to understand or dont clearly explain the column data
    - Since we are given the info of what each column represents I will not change them

***Column based info:***
- Most clients have 0 children
- There are 47 clients with -1 children, this could be a typo or meant to represent clients who have lost a child
- There is a large age range of clients
- Most clients appear to have a secondary education, the least have graduate degrees
- Most clients are married followed by civil partnership, unmarried, divorced, widow / widower
- There are almost 2x more female clients than male clients
- Most clients are employees
- 1741 clients have defaulted on a loan before
    - This makes up 8.1% of the clients
- The explanation of the data says that total_income represents monthly income, these clients appear to be very high earners
    - Maybe this is meant to represent annual income?
- The purpose subcategories have themes but there are several representing a similar reason/explanation

# Conclusion updates after code review and code changes/updates
***Changes made to data for better representation, readability, easier analysis, etc.***

If values in columns needed to be udated, a new column was created to do so because no instructions were given to modify existing column values
- children column changed to num_children to better represent the information in the column
- days_employed values changed from negative to positive
- New column (defaulted_before) added to be more readable and better represent the debt column

***Interesting artifacts***

- The num_children column contains clients with -1 and 20 children.
    - While a -1 and 20 may be inappropriate values for number of children, with out being able to ask the investigators clarifying questions about the data and collection/transfer methods I will not change them.
       - -1 children may represent clients who have lost children
       - 20 children may be a large number of children, but is not impossible
- days_employed_pos column is meant to represent how long the customer has been working, while the column name implies number of days, some of these values are too large for an average human lifespan, perhaps they are meant to represent hours or there was a calculation error.

***Information after updates***

Initial data started with 12 columns, after updates there are 13

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: Such an amazing conclusion!
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: It is still such an amazing conclusion!
</div>

## Data preprocessing

### Processing missing values

***Process missing values in the data***

- Determine if missing values across columns are in the same rows
- See if there is a pattern to explain the missing values
- Evaluate the presence of missing values in the data
- Determine appropriate data to use to fill in missing values
- Fill in missing values
- Confirm missing values are filled
- Confirm data is not significantly impacted by filled in missing values

In [26]:
# Moved to earlier cell after code review
# read and save working version of data
# scoring_credit = pd.read_csv('/datasets/credit_scoring_eng.csv')

***Determine if missing values across columns are in the same rows:***

In [27]:
# Determine if missing values in days_employed and total_income are in the same rows
# If so, expect a return of null values ([nan])
uniq_mis_vals = scoring_credit[scoring_credit['days_employed'].isnull()]['total_income'].unique()
print('Unique missing values in days_employed and total_income columns, expect [nan]:\n{}'.format(uniq_mis_vals), '\n')

Unique missing values in days_employed and total_income columns, expect [nan]:
[nan] 



In [28]:
# Determine if the length of the returned list matches the number of .isnull() determined in previous cell (2174)
num_uniq_mis_vals = len(scoring_credit[scoring_credit['days_employed'].isnull()]['total_income'])
print('Number of unique missing values in days_employed and total_income columns, expect 2174:\n{}'.format(num_uniq_mis_vals), '\n')


Number of unique missing values in days_employed and total_income columns, expect 2174:
2174 



The returned values above indicate that the missing values in the days_employed and total_income columns are in the same rows.

***Review the rows to determine if there is an obvious pattern as to why the values are missing:***

In [29]:
# Review rows with missing values to see if there is an obvious pattern
print('First 20 rows with missing data:')
print(scoring_credit[scoring_credit['days_employed'].isnull()].head(20), '\n') # no obvious pattern - let's take a closer look

First 20 rows with missing data:
     num_children  days_employed  dob_years            education  \
12              0            NaN         65  secondary education   
26              0            NaN         41  secondary education   
29              0            NaN         63  secondary education   
41              0            NaN         50  secondary education   
55              0            NaN         54  secondary education   
65              0            NaN         21  secondary education   
67              0            NaN         52    bachelor's degree   
72              1            NaN         32    bachelor's degree   
82              2            NaN         50    bachelor's degree   
83              0            NaN         52  secondary education   
90              2            NaN         35    bachelor's degree   
94              1            NaN         34    bachelor's degree   
96              0            NaN         44  SECONDARY EDUCATION   
97             

Reviewing the first 20 rows of the data does not reveal an obvious pattern in rows with missing values.

***Evaluate the subcategories of other columns to determine if there is a pattern associated with the missing values.***

The investigation is asking for possible reasons for loan defaults. Because several of the columns have data that still need to be cleaned up family_status seems like a good place to start as it is a focus for the investigation and has clean data

In [30]:
# Evaluate the subcategories assocatied with missing data in the family_status column
fam_stat_sub_cat_mis_val_count = scoring_credit[scoring_credit['days_employed'].isnull()]['family_status'].value_counts()
print('Count of family_status subcategories associated with missing values:\n{}'
      .format(fam_stat_sub_cat_mis_val_count), '\n')

Count of family_status subcategories associated with missing values:
married              1237
civil partnership     442
unmarried             288
divorced              112
widow / widower        95
Name: family_status, dtype: int64 



In [31]:
# Confirm that the number of counts associated with missing values totals to the number of missing information
fam_stat_sub_cat_mis_val_total = scoring_credit[scoring_credit['days_employed'].isnull()]['family_status'].value_counts().sum()
print('Total number of family_status rows with missing values (expect 2174):\n{}'
      .format(fam_stat_sub_cat_mis_val_total), '\n')

Total number of family_status rows with missing values (expect 2174):
2174 



Evaluate the subcategory conversion rates:

In [32]:
# Evaluate the subcategory converstion rates:
# Compared to total of each subcategories for family_status column
fam_stat_sub_cat_counts = scoring_credit['family_status'].value_counts()
fam_stat_sub_cat_comp_mis_total = fam_stat_sub_cat_mis_val_count / fam_stat_sub_cat_counts
print('Conversion rate of missing values associated with each family_status subcategory:\n{}'
      .format(fam_stat_sub_cat_comp_mis_total), '\n')

Conversion rate of missing values associated with each family_status subcategory:
married              0.099919
civil partnership    0.105818
unmarried            0.102382
divorced             0.093724
widow / widower      0.098958
Name: family_status, dtype: float64 



In [33]:
# Conversion rate compared to total entries
fam_stat_sub_cat_mis_val_con_rate = scoring_credit[scoring_credit['days_employed'].isnull()]['family_status'].value_counts()
print('Conversion rate of missing values associated with each family_status subcategory in total entries:\n{}'
      .format(fam_stat_sub_cat_mis_val_con_rate / len(scoring_credit)), '\n')

Conversion rate of missing values associated with each family_status subcategory in total entries:
married              0.057468
civil partnership    0.020534
unmarried            0.013380
divorced             0.005203
widow / widower      0.004413
Name: family_status, dtype: float64 



While 10% of each family_status subcategory is associated with missing values, there does not seem to be a specific pattern as to why.

Although the conversion rates for each family_status subcategory is 6% or less of the total entries, the directions given are to fill in missing information rather than delete them.

***Determine the appropriate data to use to fill missing values.***

Calculate mean and median of columns with missing values

In [34]:
# Calculate mean and median of columns with missing values
days_employed_mean = scoring_credit['days_employed'].mean()
days_employed_median = scoring_credit['days_employed'].median()
total_income_mean = scoring_credit['total_income'].mean()
total_income_median = scoring_credit['total_income'].median()
print('days_employed mean: {}'.format(days_employed_mean))
print('days_employed median: {}'.format(days_employed_median), '\n')
print('total_income mean: {}'.format(total_income_mean))
print('total_income median: {}'.format(total_income_median), '\n')

days_employed mean: 66914.72890682236
days_employed median: 2194.220566878695 

total_income mean: 26787.56835465867
total_income median: 23202.87 



In [35]:
# Evaluate the mean and median of the family_status subcategories
fam_stat_sub_cat_mean = scoring_credit.groupby('family_status')['days_employed', 'total_income'].mean()
print('Mean of family_status subcategories in columns with missing values:\n{}'.format(fam_stat_sub_cat_mean), '\n')
fam_stat_sub_cat_median = scoring_credit.groupby('family_status')['days_employed', 'total_income'].median()
print('Median of family_status subcategories in columns with missing values:\n{}'.format(fam_stat_sub_cat_median), '\n')

Mean of family_status subcategories in columns with missing values:
                   days_employed  total_income
family_status                                 
civil partnership   58396.260670  26694.428597
divorced            68816.335483  27189.354550
married             63312.782890  27041.784689
unmarried           47072.691647  26934.069805
widow / widower    205636.887848  22984.208556 

Median of family_status subcategories in columns with missing values:
                   days_employed  total_income
family_status                                 
civil partnership    1944.345291     23186.534
divorced             2401.954568     23515.096
married              2304.964439     23389.540
unmarried            1462.009287     23149.028
widow / widower    337017.713307     20514.190 



The widow / widower subcategory has a high median. However the median values for the other subcategories are close to the days_employed and total_income column medians.

The means are higher than the median implying that there are outliers in the data. 

I am going to use the median values of the days_employed and total_income columns to fill in the missing values. 

***Fill in missing values with out significantly impacting the analysis:***

In [36]:
# Fill in days_employed_pos missing values with median of that column
scoring_credit.loc[(scoring_credit['days_employed'].isnull()), 'days_employed'] = days_employed_median

<div class="alert alert-warning" style="box-shadow: 4px 4px 4px">
Remark: There is a typo in this code cell. Please, find it. It doesn't affect the fact that work is accepted!
</div>

In [37]:
# Fill in total_income missing values with median of that column                    
scoring_credit.loc[(scoring_credit['total_income'].isnull()), 'total_income'] = total_income_median

***Confirm the missing values have been filled in***

In [38]:
# Review data after missing values filled
print('Review first 30 rows of data after missing values filled:')
print(scoring_credit.head(30), '\n')

Review first 30 rows of data after missing values filled:
    num_children  days_employed  dob_years            education  education_id  \
0              1    8437.673028         42    bachelor's degree             0   
1              1    4024.803754         36  secondary education             1   
2              0    5623.422610         33  Secondary Education             1   
3              3    4124.747207         32  secondary education             1   
4              0  340266.072047         53  secondary education             1   
5              0     926.185831         27    bachelor's degree             0   
6              0    2879.202052         43    bachelor's degree             0   
7              0     152.779569         50  SECONDARY EDUCATION             1   
8              2    6929.865299         35    BACHELOR'S DEGREE             0   
9              0    2188.756445         41  secondary education             1   
10             2    4171.483647         36    bache

In [39]:
# Confirm that missing values are filled in
fam_stat_sub_cat_count_post_fill = scoring_credit[scoring_credit['days_employed'].isnull()]['family_status'].value_counts()
print('family_status counts associated with missing values filled (expect an empty series):\n{}'.format(fam_stat_sub_cat_count_post_fill), '\n')

family_status counts associated with missing values filled (expect an empty series):
Series([], Name: family_status, dtype: int64) 



In [40]:
# Expect a return of 0
num_uniq_mis_vals_post_fill = len(scoring_credit[scoring_credit['days_employed'].isnull()]['total_income'])
print('Number of unique missing values in days_employed and total_income columns with missing values filled (expect 0)\n{}'
      .format(num_uniq_mis_vals_post_fill), '\n')

Number of unique missing values in days_employed and total_income columns with missing values filled (expect 0)
0 



Confirm that there are no more missing values:

In [41]:
# Confirm that there are no unique values between the days_employed_pos and total_income columns with missing values filled in:
# Expect an empty list
uniq_mis_vals_post_fill = scoring_credit[scoring_credit['days_employed'].isnull()]['total_income'].unique()
print('Unique missing values in days_employed and total_income columns with missing values filled (expect an empyt list)\n{}'
      .format(uniq_mis_vals_post_fill), '\n')

Unique missing values in days_employed and total_income columns with missing values filled (expect an empyt list)
[] 



In [42]:
# Confirm all missing values are filled in the data
mis_val_post_fill = scoring_credit.isnull().sum()
print('Missing values after missing values have been filled:\n{}.'.format(mis_val_post_fill))

Missing values after missing values have been filled:
num_children        0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
defaulted_before    0
dtype: int64.


***Confirm data wasnt significanlty impacted by filling in missing values***

Examine the subcategory means and medians with missing values filled in:

In [43]:
# family_status subcategory mean with missing values filled in:
fam_stat_sub_cat_mean_post_filled = scoring_credit.groupby('family_status')['days_employed', 'total_income'].mean()
print('Mean of family_status subcategories with missing values filled:\n{}'.format(fam_stat_sub_cat_mean_post_filled), '\n')

Mean of family_status subcategories with missing values filled:
                   days_employed  total_income
family_status                                 
civil partnership   52449.097221  26324.960342
divorced            62572.254419  26815.725873
married             57205.863537  26658.203310
unmarried           42477.953051  26552.062857
widow / widower    185504.540565  23005.846928 



In [44]:
# family_status subcategory median with missing values filled in:
fam_stat_sub_cat_median_post_filled = scoring_credit.groupby('family_status')['days_employed', 'total_income'].median()
print('Median of family_status subcategories with missing values filled:\n{}'.format(fam_stat_sub_cat_median_post_filled), '\n')

Median of family_status subcategories with missing values filled:
                   days_employed  total_income
family_status                                 
civil partnership    2194.220567    23202.8700
divorced             2194.220567    23202.8700
married              2194.220567    23202.8700
unmarried            1757.478271    23202.8700
widow / widower    329131.870600    21915.4555 



In [45]:
# Mean and median of days_employed_pos and total_income columns with missing values filled in
days_employed_mean_filled = scoring_credit['days_employed'].mean()
days_employed_median_filled = scoring_credit['days_employed'].median()
total_income_mean_filled = scoring_credit['total_income'].mean()
total_income_median_filled = scoring_credit['total_income'].median()
print('days_employed mean with missing values filled: {}'.format(days_employed_mean_filled))
print('days_employed median with missing values filled: {}'.format(days_employed_median_filled), '\n')
print('total_income mean with missing values filled: {}'.format(total_income_mean_filled))
print('total_income median with missing values filled: {}'.format(total_income_median_filled), '\n')

days_employed mean with missing values filled: 60378.0327334873
days_employed median with missing values filled: 2194.220566878695 

total_income mean with missing values filled: 26425.51798425087
total_income median with missing values filled: 23202.87 



Compare above numbers to pre-filled values:

In [46]:
# Compare pre-filled missing value mean and median values:
print('Mean and median before missing values were filled:')
print('days_employed mean: {}'.format(days_employed_mean))
print('days_employed median: {}'.format(days_employed_median), '\n')
print('total_income mean: {}'.format(total_income_mean))
print('total_income median: {}'.format(total_income_median), '\n')

Mean and median before missing values were filled:
days_employed mean: 66914.72890682236
days_employed median: 2194.220566878695 

total_income mean: 26787.56835465867
total_income median: 23202.87 



days_employed mean went up but total_income mean and median are similar.

In [47]:
# Personal curiosity of if rows with missing values were dropped instead of filled in
#df_dropna = pd.read_csv('/datasets/credit_scoring_eng.csv')
#df_dropna.dropna().reset_index(drop=True)
#print('days_employed mean dropped missing values:', df_dropna['days_employed'].mean())
# print('days_employed median dropped missing values:', df_dropna['days_employed'].median())
# print('total_income mean dropped missing values:', df_dropna['total_income'].mean())
# print('total_income median dropped missing values:', df_dropna['total_income'].median())

<div class="alert alert-danger" style="box-shadow: 4px 4px 4px">
<s>Need fixing: This part of code is correct. However, it is inappropriate for a real report. Please, make only one output in one code cell. Also, left your comments in markdowns. I will not write this for subsequent cells, but please correct it for each cell.</s>
</div>

### Conclusion

***Location of the missing values:***
- Missing values in the days_employed_pos and total_income colmns are in the same rows
    - Both are float data types
- There is no obvious pattern in the column subgroups for the missing data it seems to be MCAR
    - Without knowing the collection method, I can only speculate why the values are missing
    - Possible reasons for missing values:
        - Providing this information is optional
        - The information was incorrectly entered
        - Data transfer error

***Filling in the missing values:***
- The instructions given say to fill in missing values vs delete them. Several columns still need to be cleaned up or arent appropriate for evaluating the missing values. I worked with the family_status column because the data is clean and is a focus of the investigation
- ***updated sentence above post code review and code updates*** Several columns still need to be cleaned up or arent appropriate for evaluating the missing values. I worked with the family_status column because the data is clean and is a focus of the investigation
    - Each subcategory had roughly 10% with missing values. While the subcategories with missing values made up 6% or less of total entries
    - Again no obvious pattern for the missing values
    
To fill in the missing values the mean and medians were evaluated:
- The mean for the columns with missing values is higher than the medians suggesting that there are outliers in both columns
- The means in the family_status subcategories in columns with missing values were lower than the corresponding column with missing values
- The medians in the family_status subcategories in columns with missing values were roughly the same as the corresponding column with missing values with the exception of the widow / widower subcategory. Which was very high for the days_employed  median and low for the total_income median.
- Missing values were filled in with the median of that column because it is unlikely to significantly impact the analysis - only 10% of the the total entries have missing values.

***Data after missing values were filled:***
- The mean of the days_employed column went down a little bit, but the total_income mean stayed about the same
- ***updated sentence above post code review and code updates*** The mean of the days_employed column went up, but the total_income mean stayed about the same
- Filling in the missing values did change the means of the family_status subcategories, suggesting that other column's subcategories would be affected by this change too. This change was made to 10% of total entries and is not likely to significantly impact the analysis

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: I couldn't agree more! 
</div>

### Data type replacement

***Convert float data type to int data type***

- Two columns need to be changed to int data type
- Use .astype() to convert a float data type to int data type
- Confirm that data types are changed

***Convert to int data type***

In [48]:
# Convert float data type to int data type:

# Use .astype() method to convert float64 type to int64 type vs to_numeric which converts strings to floats
scoring_credit['days_employed'] = scoring_credit['days_employed'].astype(int)

scoring_credit['total_income'] = scoring_credit['total_income'].astype(int)

***Confirm data types are changed***

In [49]:
# Confirm the data types have changed
print('Dataframe info with updated data types:')
scoring_credit.info()

Dataframe info with updated data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
num_children        21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
defaulted_before    21525 non-null object
dtypes: int64(7), object(6)
memory usage: 2.1+ MB


In [50]:
print('First 10 rows of dataframe after data types were updated:')
print(scoring_credit.head(10))

First 10 rows of dataframe after data types were updated:
   num_children  days_employed  dob_years            education  education_id  \
0             1           8437         42    bachelor's degree             0   
1             1           4024         36  secondary education             1   
2             0           5623         33  Secondary Education             1   
3             3           4124         32  secondary education             1   
4             0         340266         53  secondary education             1   
5             0            926         27    bachelor's degree             0   
6             0           2879         43    bachelor's degree             0   
7             0            152         50  SECONDARY EDUCATION             1   
8             2           6929         35    BACHELOR'S DEGREE             0   
9             0           2188         41  secondary education             1   

       family_status  family_status_id gender income_type  de

### Conclusion

***Updating data types - float to int:***
- Used .astype(int) method to convert float64 type to int64 type because float64 is a numeric data type.
- If the initial data type was a string, then .to_numeric() would be used which converts strings to floats and then .astype() to convert the float to int.

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: This part is flawless!
</div>

### Processing duplicates

***Remove duplicate rows***

It was determined earlier that the education column contains case discrepencies. This needs to be resolved before the duplicates can be removed

- Review number of duplicates after data has been updated
- Resolve case discrepencies in the education column
    - Change column values to lower case
- Check number of duplicates after column update
- Examine duplicates for a pattern to explain reason for duplicates 
- Remove dupliates
- Confirm removal of duplicates

In [51]:
# Review number of duplicates after missing values filled and data types changed
sum_df_dups = scoring_credit.duplicated().sum()
print('Number of duplicate rows before education column is converted to lower case:\n{}'.format(sum_df_dups), '\n')

Number of duplicate rows before education column is converted to lower case:
54 



In [52]:
# Update the education column with lower case letters
scoring_credit['education'] = scoring_credit['education'].str.lower()
print(scoring_credit.head(10), '\n')

   num_children  days_employed  dob_years            education  education_id  \
0             1           8437         42    bachelor's degree             0   
1             1           4024         36  secondary education             1   
2             0           5623         33  secondary education             1   
3             3           4124         32  secondary education             1   
4             0         340266         53  secondary education             1   
5             0            926         27    bachelor's degree             0   
6             0           2879         43    bachelor's degree             0   
7             0            152         50  secondary education             1   
8             2           6929         35    bachelor's degree             0   
9             0           2188         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0

In [53]:
# Determine number of duplicates after education column is converted to lower case
sum_df_dups_lower = scoring_credit.duplicated().sum()
print('Number of duplicate rows after education column converted to lower case:\n{}'.format(sum_df_dups_lower), '\n')

Number of duplicate rows after education column converted to lower case:
71 



***Investigate the duplicates to determine if there is a pattern or cause for them***

In [54]:
# Calculate conversion rate of duplicates in total entries
dup_conv_rate = scoring_credit.duplicated().sum() / len(scoring_credit)
print('Percent of duplicates in total entries:\n{:0.1%}'.format(dup_conv_rate), '\n')

Percent of duplicates in total entries:
0.3% 



In [55]:
# Create a new dataframe with only the duplicate rows
dup_rows = scoring_credit[scoring_credit.duplicated()]
print('Dataframe of duplicates only information:')
dup_rows.info()

Dataframe of duplicates only information:
<class 'pandas.core.frame.DataFrame'>
Int64Index: 71 entries, 2849 to 21415
Data columns (total 13 columns):
num_children        71 non-null int64
days_employed       71 non-null int64
dob_years           71 non-null int64
education           71 non-null object
education_id        71 non-null int64
family_status       71 non-null object
family_status_id    71 non-null int64
gender              71 non-null object
income_type         71 non-null object
debt                71 non-null int64
total_income        71 non-null int64
purpose             71 non-null object
defaulted_before    71 non-null object
dtypes: int64(7), object(6)
memory usage: 7.8+ KB


The data below represents the duplicates after the first record, i.e. it is showing the second and/or third records

In [56]:
print('First 20 rows of duplicates only data frame:\n')
print(dup_rows.head(20), '\n') 

First 20 rows of duplicates only data frame:

       num_children  days_employed  dob_years            education  \
2849              0           2194         41  secondary education   
3290              0           2194         58  secondary education   
4182              1           2194         34    bachelor's degree   
4851              0           2194         60  secondary education   
5557              0           2194         58  secondary education   
6312              0           2194         30  secondary education   
7808              0           2194         57  secondary education   
7921              0           2194         64    bachelor's degree   
7938              0           2194         71  secondary education   
8583              0           2194         58    bachelor's degree   
9238              2           2194         34  secondary education   
9528              0           2194         66  secondary education   
9604              0           2194         7

The percentage of duplicates is small: 0.3%

Based on the first 20 rows:
- There is no obvious pattern
- Duplicates appear to have contained missing values in the days_employed_pos and total_income columns
    - These columns contain the median determined above
    
***Examine the subcategories of the duplicate rows to look for a pattern or cause of duplicates:***

In [57]:
# Take a closer look at the columns to look for a pattern or cause for duplicates
fam_stat_dups = dup_rows['family_status'].value_counts()
print('family_status subcategories with duplicates:\n{}'.format(fam_stat_dups), '\n')

family_status subcategories with duplicates:
married              41
civil partnership    26
unmarried             3
widow / widower       1
Name: family_status, dtype: int64 



In [58]:
dob_years_dups = dup_rows['dob_years'].value_counts()
print('dob_years subcategories with duplicates:\n{}'.format(dob_years_dups), '\n')

dob_years subcategories with duplicates:
58    7
64    5
62    4
57    4
56    4
47    3
30    3
60    3
54    3
46    3
71    2
34    2
40    2
41    2
44    2
23    2
48    2
51    2
36    1
29    1
65    1
31    1
32    1
61    1
35    1
37    1
66    1
38    1
39    1
59    1
42    1
43    1
50    1
45    1
Name: dob_years, dtype: int64 



In [59]:
education_dups = dup_rows['education'].value_counts()
print('education subcategories with duplicates:\n{}'.format(education_dups), '\n')

education subcategories with duplicates:
secondary education    61
bachelor's degree      10
Name: education, dtype: int64 



In [60]:
income_type_dups = dup_rows['income_type'].value_counts()
print('income_type subcategories with duplicates:\n{}'.format(income_type_dups), '\n')

income_type subcategories with duplicates:
employee         35
retiree          27
business          7
civil servant     2
Name: income_type, dtype: int64 



In [61]:
purpose_dups = dup_rows['purpose'].value_counts()
print('purpose subcategories with duplicates:\n{}'.format(purpose_dups), '\n')

purpose subcategories with duplicates:
having a wedding                            9
to have a wedding                           9
wedding ceremony                            6
housing renovation                          5
to become educated                          4
transactions with my real estate            3
buy real estate                             3
second-hand car purchase                    3
buy commercial real estate                  3
purchase of the house for my family         3
buying property for renting out             2
building a real estate                      2
supplementary education                     2
to own a car                                2
housing                                     1
to buy a car                                1
building a property                         1
to get a supplementary education            1
purchase of the house                       1
real estate transactions                    1
buy residential real estate              

No obvious pattern seems to be present

Since there is no unique identifier for the information, it seems easy to accidentally make a copy, or perhaps the client filled out information twice

***Remove duplicate rows***

In [62]:
# Drop the duplilcates and reset index
scoring_credit = scoring_credit.drop_duplicates().reset_index(drop=True)

***Confirm duplicates were removed***

In [63]:
# Check data
print('Data information after duplicates are dropped:')
scoring_credit.info()

Data information after duplicates are dropped:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
num_children        21454 non-null int64
days_employed       21454 non-null int64
dob_years           21454 non-null int64
education           21454 non-null object
education_id        21454 non-null int64
family_status       21454 non-null object
family_status_id    21454 non-null int64
gender              21454 non-null object
income_type         21454 non-null object
debt                21454 non-null int64
total_income        21454 non-null int64
purpose             21454 non-null object
defaulted_before    21454 non-null object
dtypes: int64(7), object(6)
memory usage: 2.1+ MB


In [64]:
# Confirm the duplicates were dropped
sum_df_dups_postdrop = scoring_credit.duplicated().sum()
print('Number of duplicates after duplicates are dropped (expect 0):\n{}'.format(sum_df_dups_postdrop))

Number of duplicates after duplicates are dropped (expect 0):
0


### Conclusion

***Looking at the duplicate rows:***
- The data contained 54 duplicates which made up 0.3% of total entries.
- ***updated sentence above post code review and code updates*** Initially the data contained 54 duplicates. After the education columns case descrepencies were resolved there were 71 duplicates making up 0.3% of the total entries
- The sum of the duplicates were found using the .duplicated().sum() methods.
- The .duplicated() method was used because the whole dataframe was being evaluated for duplicates vs a column, in which case value_counts() may be more appropriate.
- A duplicate only dataframe was created with .duplicates() as the logic.
- Several subcategories within the duplicate only dataframe were examined to see if there is a pattern as to why they occured. 
    - Most of the rows appear to have been in the missing values rows because the days_employed_pos and total_income contain the median value previously filled in.
    - The amounts of duplicates in each subcategory appear to be representative of the total of their respective subcategories. Since a pattern was not obvious this wasnt evaluated mathematically.
    - Possible reasons for duplicates:
        - Data transfer error
        - Clients may have submitted information twice - there is no unique identifier for each client, so they wouldnt be able to check if they have submitted information already
        - Technical error while clients were filling submitting information may have made it seem like their information didnt go through
        
***Deleting in the duplicate rows:***
- .drop_duplicates() method was used to on the whole dataframe to remove duplicates because this will remove all duplicate rows.

***Data after duplicate rows deleted:***
- Number of entries shifted to 21471
- ***updated sentence above post code review and code updates*** Number of entries shifted to 21454
- 0 duplicates found after duplicates were removed

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: Again, this part is perfect. However, it is really inappropriate for a real report to wite the whole step in one code cell.  
</div>

### Categorizing Data

Some of the columns have values that are mostly unique, represent similar things in a different way, or are too broad to evaluate easily and need to be categorized to make evaluation easier.
- The purpose column has different phrases representing the same thing
    - Join the words from the purpose column and tokenize them into a list of strings
    - Use lemmatization to determine the main focus from the phrases
    - Search for the stem of the words
    - Use the stems to make a new column containing the focus of the loan use
- The total_income column has a broad range of values
    - Group income levels in groups of 25,000
    - Create a new column to represent the groups of income levels
    - Examine the results to determine if these groups are appropriate
- Confirm that all values were processed into new columns and are as expected

***Join all the words from the purpose column and tokenize them***

In [65]:
# Join all words from purpose column
allwords = ' '.join(scoring_credit['purpose'])
# print('Data type of all words from purpose column combined:')
# print(type(allwords), '\n')
# Make sure all the words are lower case
allwords = allwords.lower()
# Tokenize all the words before lemmatization
wordnet_lemma = WordNetLemmatizer()
allwords = nltk.word_tokenize(allwords)
# confirm allwords is a list of strings
print('First 10 words of all words from column combined after being tokenized, expect a list of strings:')
print(allwords[:10], '\n') 

First 10 words of all words from column combined after being tokenized, expect a list of strings:
['purchase', 'of', 'the', 'house', 'car', 'purchase', 'purchase', 'of', 'the', 'house'] 



***Lemmatize the words and use the appropriate nouns as stems***

In [66]:
lemmas = [wordnet_lemma.lemmatize(w, pos = 'n') for w in allwords]

# Count the number of lemmas from the purpose column
print('Count of all words used in purpose column:')
print(Counter(lemmas), '\n')

Count of all words used in purpose column:
Counter({'a': 5104, 'real': 4464, 'estate': 4464, 'car': 4306, 'purchase': 3306, 'education': 3109, 'to': 3064, 'of': 2994, 'transaction': 2604, 'property': 2538, 'my': 2390, 'buy': 2359, 'wedding': 2324, 'own': 2238, 'housing': 1905, 'house': 1904, 'buying': 1634, 'commercial': 1311, 'for': 1289, 'the': 1284, 'with': 1277, 'building': 1243, 'second-hand': 964, 'university': 948, 'supplementary': 906, 'getting': 868, 'ceremony': 791, 'having': 768, 'have': 765, 'renting': 651, 'out': 651, 'family': 638, 'construction': 635, 'renovation': 607, 'residential': 606, 'going': 496, 'get': 446, 'an': 442, 'profile': 436, 'higher': 426, 'become': 408, 'educated': 408}) 



In [67]:
# Create a list of words pulled from purpose column that focus on the main use for the loan
pre_stem_list = ['estate', 'car', 'education', 'property', 'wedding', 'housing', 'house', 'educated', 'university']
print('List of words from purpose column to break down into stems:\n{}'.format(pre_stem_list), '\n')

List of words from purpose column to break down into stems:
['estate', 'car', 'education', 'property', 'wedding', 'housing', 'house', 'educated', 'university'] 



In [68]:
# Set stemmer to english
english_stemmer = SnowballStemmer('english')

# Save stems of the words in a list and print them
stems = []
for word in pre_stem_list:
    stem_word = english_stemmer.stem(word)
    stems.append(stem_word)
    
print('Stem words used for loan use from purpose column:')
print(stems, '\n')

Stem words used for loan use from purpose column:
['estat', 'car', 'educ', 'properti', 'wed', 'hous', 'hous', 'educ', 'univers'] 



- The stem for housing and house is hous 
- The stem for education and educated is educ

This will be helpful when looking for and categorizing these words in the purpose column

***Use the stems to categorize the main purpose of the loan***

In [69]:
# Use stems to categorize the purpose column into more specific subcategories
# Create variables for each stem
estate_stem = 'estat'
car_stem = 'car'
educate_stem = 'educ'
property_stem = 'properti'
wedding_stem = 'wed'
house_stem = 'hous'
university_stem = 'univers'

# Create a function to determine the main use of the loan
def assign_loan_use(purpose):
    # Create a default subcategory
    loan_use = 'unknown loan use'
    # Tokenize the words of each row in purpose column
    tok_words = nltk.word_tokenize(purpose)
    # Iterate over the words (now a list of strings) from each row in the purpose column
    for word in tok_words:
        if english_stemmer.stem(word) == estate_stem:
            loan_use = 'real estate'
        if english_stemmer.stem(word) == car_stem:
            loan_use = 'car'
        if english_stemmer.stem(word) == educate_stem or english_stemmer.stem(word) == university_stem:
            loan_use = 'education'
        if english_stemmer.stem(word) == property_stem:
            loan_use = 'property'
        if english_stemmer.stem(word) == wedding_stem:
            loan_use = 'wedding'
        if english_stemmer.stem(word) == house_stem:
            loan_use = 'house'
    return loan_use

***Create a new column in the dataframe categorizing the main use of the loan***

In [70]:
# Create a new column that represents the main use of the loan as described in the pupose column:
# Use the .apply() method to apply the purpose column to the assign_loan_use function
scoring_credit['loan_use'] = scoring_credit['purpose'].apply(assign_loan_use)

***Examine the total_income column***

In [71]:
# Determine min and max of total income
# mean and median were previously determined: 
# total_income mean with missing values filled: 26425.51798425087
# total_income median with missing values filled: 23202.87 
total_income_max = scoring_credit['total_income'].max()
total_income_min = scoring_credit['total_income'].min()
print('total_income max:\n{}'.format(total_income_max))
print('total_income min:\n{}'.format(total_income_min), '\n')

total_income max:
362496
total_income min:
3306 



Group the data by ranges of 25,000: 

In [72]:
# Create a function that categorizes total_income into ranges of income
def assign_income_range(income):
    if income < 25000:
        return '$0 - $24,999'
    elif 25000 <= income < 50000:
        return '$25,000 - $49,999'
    elif 50000 <= income < 75000:
        return '$50,000 - $74,999'
    elif 75000 <= income < 100000:
        return '$75,000 - $99,999'
    elif 100000 <= income < 125000:
        return '$100,000 - $124,999'
    elif 12500 <= income < 150000:
        return '$125,000 - $149,999'
    elif income >= 150000:
        return '> $150,000'
    else:
        return 'income range unknown'

***Create a new column in the dataframe categorizing the income levels***

In [73]:
# Create a new column that categorizes total_income into ranges:
# Use the .apply() method to apply the total_income column to the function assign_income_range
scoring_credit['income_range'] = scoring_credit['total_income'].apply(assign_income_range)

***Examine updated dataframe***

In [74]:
# Print updated dataframe
print('First 10 rows of dataframe with new columns:')
print(scoring_credit.head(10),'\n')

First 10 rows of dataframe with new columns:
   num_children  days_employed  dob_years            education  education_id  \
0             1           8437         42    bachelor's degree             0   
1             1           4024         36  secondary education             1   
2             0           5623         33  secondary education             1   
3             3           4124         32  secondary education             1   
4             0         340266         53  secondary education             1   
5             0            926         27    bachelor's degree             0   
6             0           2879         43    bachelor's degree             0   
7             0            152         50  secondary education             1   
8             2           6929         35    bachelor's degree             0   
9             0           2188         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_inc

***Confirm all rows were processed as expected***

In [75]:
# Confirm that all purpose rows were processed with the apply_loan_use function
loan_use_total = scoring_credit['loan_use'].count()
print('loan_use total (expect 21454):\n{}'.format(loan_use_total), '\n')

loan_use total (expect 21454):
21454 



In [76]:
loan_use_sub_cat_totals = scoring_credit['loan_use'].value_counts()
print('loan_use subcategories:\n{}'.format(loan_use_sub_cat_totals), '\n')

loan_use subcategories:
real estate    4464
car            4306
education      4013
house          3809
property       2538
wedding        2324
Name: loan_use, dtype: int64 



In [77]:
# Confirm that all income_level rows were processed with the apply_income_range function
income_range_total = scoring_credit['income_range'].count()
print('income_range total (expect 21454):\n{}'.format(income_range_total), '\n')

income_range total (expect 21454):
21454 



In [78]:
# Evaluate counts of income ranges to see if created ranges are appropriate
income_range_sub_cat_totals = scoring_credit['income_range'].value_counts()
print('Income range subcategories:\n{}'.format(income_range_sub_cat_totals), '\n')

Income range subcategories:
$0 - $24,999           12851
$25,000 - $49,999       7283
$50,000 - $74,999       1032
$75,000 - $99,999        189
$100,000 - $124,999       55
> $150,000                28
$125,000 - $149,999       16
Name: income_range, dtype: int64 



<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: Great that lematization was used!
</div>

### Conclusion

***Data to categorize:***
- Categorization is useful because it groups data that are individual peices into general categories that are easier to analyze.
- The investigation calls for an evaluation of number of children, marital status, income level, and loan purpose.
    - The children and family_status columns are already well categorized.
    - total_income and purpose columns were categorized in this analysis.
    
***Categorizing the data:***
- total_income column was categorized into groups of 25,000 until 149,999, then > 150,000.
    - This grouping appears to be an appropriate way to subcategorize the total_income column.
- purpose column was categorized into groups derived from selected nouns in the entry given by the client.
    - This lead to 6 different subcategories.
        - While real estate, house, and property could be grouped together, I left them separate because I believe they represent differnt things.

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: Absolutely correct! 
</div>

<div class="alert alert-warning" style="box-shadow: 4px 4px 4px">
Remark: Why did you decide not to change the values 20 and -1 in the children column and negative values in days_employed? It is not necessary to change, but it is but it is desirable to explain why you do not change it. Obviously, it affect the whole analysis.  
</div>

## Answer these questions

- Is there a relation between having kids and repaying a loan on time?

The contents in the cell below was moved to more appropriate locations after code review and code updates

In [79]:
# import numpy as np

# # Rename children column to be more understandable
# # scoring_credit.rename(columns={'children': 'num_children'}, inplace=True)

# # Create a more readable column for those that have defaulted on a loan before:
# # Create a function to change the column values
# def assign_default_status(default_status):
#     if default_status == 1:
#         return 'Yes'
#     if default_status == 0:
#         return 'No'
#     else:
#         return 'Unknown'

# # Create new column with newly assigned default history
# scoring_credit['defaulted_before'] = scoring_credit['debt'].apply(assign_default_status)
# # Check data
# print(scoring_credit.head(), '\n')
# print('defaulted_before subcategory totals:')
# print(scoring_credit['defaulted_before'].value_counts(), '\n')

***Create a pivot table showing relationship between having kids and repaying a loan on time***

- Calculate the subcategory representation in total entries
- Calculate the chance of defaulting on a loan according to this data set

In [80]:
# Create a pivot table showing relationship between having kids and repaying a loan on time
kids_and_loans_table = scoring_credit.pivot_table(index='num_children', columns='defaulted_before', 
                                                  values='debt', aggfunc='count', fill_value=0, margins=True)

In [81]:
# make a column for subcategory representation in all entries
kids_and_loans_table['% in tot entries'] = (kids_and_loans_table['All'] / len(scoring_credit) * 100).round(2)

# if you have x num of children you have a _% chance of defaulting
kids_and_loans_table['% chance of default'] = (kids_and_loans_table['Yes'] / kids_and_loans_table['All'] * 100).round(1)

print('Pivot table showing relationship between number of children and repaying a loan on time:\n')
print(kids_and_loans_table.head(10))

Pivot table showing relationship between number of children and repaying a loan on time:

defaulted_before     No   Yes    All  % in tot entries  % chance of default
num_children                                                               
-1                   46     1     47              0.22                  2.1
0                 13028  1063  14091             65.68                  7.5
1                  4364   444   4808             22.41                  9.2
2                  1858   194   2052              9.56                  9.5
3                   303    27    330              1.54                  8.2
4                    37     4     41              0.19                  9.8
5                     9     0      9              0.04                  0.0
20                   68     8     76              0.35                 10.5
All               19713  1741  21454            100.00                  8.1


<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
    Success: Great that function and <code>.apply()</code> were used!
</div>

### Conclusion

Most clients have 0 children, followed by 1, then 2.

- If -1 number of children is not a typo, then these clients have the lowest chance of defaulting (2.1%).
- If -1 number of children is a typo, then clients with 0 children have the lowest chance of defaulting (7.5%).
- Clients with 20 children have the highest chance of defaulting (10.5%).

According to this data set, there is a trend suggesting that based on the number of children a client has, i.e. the more children, the higher the chance of defaulting on a loan.

Possible reasons for these results are that the more children a client has, the higher the financial strain, which can lead to conflicting priorities for paying a loan on time.

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: Perfect conclusion!
</div>

- Is there a relation between marital status and repaying a loan on time?

***Create a pivot table showing relationship between marital status and repaying a loan on time***

- Calculate the subcategory representation in total entries
- Calculate the chance of defaulting on a loan according to this data set

In [82]:
# Create a pivot table showing the relationship between marital status and repaying a loan on time
fam_stat_and_loans_table = scoring_credit.pivot_table(index='family_status', columns='defaulted_before',
                                                      values='debt', aggfunc='count', margins=True)

In [83]:
print('Pivot table showing relationship between marital status and repaying a loan on time:\n')

# make a column for subcategory representation in all entries
fam_stat_and_loans_table['% in tot entries'] = (fam_stat_and_loans_table['All'] / len(scoring_credit) * 100).round(1)

# if you are in this type of family status you have a _% chance of defaulting
fam_stat_and_loans_table['% chance of default'] = (fam_stat_and_loans_table['Yes'] / fam_stat_and_loans_table['All'] * 100).round(1)
# if you are in this type of family status you are _ times more likely to default
# fam_stat_and_loans_table['times more likely to default'] = (fam_stat_and_loans_table['No'] / fam_stat_and_loans_table['Yes']).round(1)

print(fam_stat_and_loans_table.head(10))

Pivot table showing relationship between marital status and repaying a loan on time:

defaulted_before      No   Yes    All  % in tot entries  % chance of default
family_status                                                               
civil partnership   3763   388   4151              19.3                  9.3
divorced            1110    85   1195               5.6                  7.1
married            11408   931  12339              57.5                  7.5
unmarried           2536   274   2810              13.1                  9.8
widow / widower      896    63    959               4.5                  6.6
All                19713  1741  21454             100.0                  8.1


<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: Great that pivot table was used!
</div>

<div class="alert alert-warning" style="box-shadow: 4px 4px 4px">
    Remark: Please, don't print out a dataset with print. It is more common to use <code>df</code> or <code>df.head()</code>
</div>

### Conclusion

Most clients are married, followed by civil partnership, unmarried, divorced, then widow/widower.

Widow/widower have the lowest chance of defaulting (6.6%).

Divorced and married clients have about the same chance of defaulting (7.5 & 7.1%) with a higher chance than widows/widowers (6.6%), and a lower chance than civil partnership and unmarried clients (9.3 & 9.8%).

Civil partnerships and unmarried clients have about the same and the highest chance of defaulting (9.3 & 9.8%).

There isnt a clear pattern here. This data may be inappropriate as a reason for defaulting on a loan.

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: Agree
</div>

- Is there a relation between income level and repaying a loan on time?

***Create a pivot table showing relationship between income level and repaying a loan on time***

- Calculate the subcategory representation in total entries
- Calculate the chance of defaulting on a loan according to this data set

In [84]:
# print(scoring_credit.groupby('income_range')['debt'].value_counts())

# Create a pivot table showing the relationship between income level and repaying a loan on time
income_level_and_loans_table = scoring_credit.pivot_table(index='income_range', columns='defaulted_before',
                                                      values='debt', aggfunc='count', margins=True)

In [85]:
print('Pivot table showing relationship between income level and repaying a loan on time:\n')

# make a column for subcategory representation in all entries
income_level_and_loans_table['% in tot entries'] = (income_level_and_loans_table['All'] / len(scoring_credit) * 100).round(1)

# if you are in this income level you have a _% chance of defaulting
income_level_and_loans_table['% chance of default'] = (income_level_and_loans_table['Yes'] / income_level_and_loans_table['All'] * 100).round(1)
# if you are in this income level you are _ times more likely to default
# income_level_and_loans_table['times more likely to default'] = (income_level_and_loans_table['No'] / income_level_and_loans_table['Yes']).round(1)


print(income_level_and_loans_table.head(10))

Pivot table showing relationship between income level and repaying a loan on time:

defaulted_before        No   Yes    All  % in tot entries  % chance of default
income_range                                                                  
$0 - $24,999         11785  1066  12851              59.9                  8.3
$100,000 - $124,999     53     2     55               0.3                  3.6
$125,000 - $149,999     14     2     16               0.1                 12.5
$25,000 - $49,999     6700   583   7283              33.9                  8.0
$50,000 - $74,999      956    76   1032               4.8                  7.4
$75,000 - $99,999      179    10    189               0.9                  5.3
> $150,000              26     2     28               0.1                  7.1
All                  19713  1741  21454             100.0                  8.1


### Conclusion

Most clients earn between 0-24,999 follwed by 25,000-49,999.

Clients earning between 0-24,999 and 25,000-49,999 have about the same chance of defaulting (8.0 & 8.3%). Which is higher than those earning between 50,000-74,999 and >150,000 (7.4 & 7.1%) and lower than those earning between 125,000-149,000 (12.5%).

Clients earning between 50,000-74,999 and >150,000 have about the same chance of defaulting (7.4 & 7.1%). Which is higher than those earning between 75,000-99,999 (5.3%) and lower than those earning between 0-24,999 and 25,000-49,999 (8.0 & 8.3%).

Clients earning between 100,000-124,999 have the lowest chance of defaulting (3.6%).

Clients earning between 125,000-149,000 have the highest chance of defaulting (12.5%).

Clients earning between 75,000-99,999 have a lower chance of defaulting (5.3%) than those earning between 50,000-74,999 and >150,000 (7.4 & 7.1%), but a higher chance than those earning between 100,000-124,999 (3.6%).

As income increases from 0 to 125,000 the chance of defaulting goes down, until income reaches 125,000 then it goes up.
Possible reasons for this could be because as people earn more money they tend to spend more and buy more expensive things.

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: Absolutely correct!
</div>

- How do different loan purposes affect on-time repayment of the loan?

***Create a pivot table showing relationship between loan purpose and repaying a loan on time***

- Calculate the subcategory representation in total entries
- Calculate the chance of defaulting on a loan according to this data set

In [86]:
#print(scoring_credit.groupby('loan_use')['debt'].value_counts())

# Create a pivot table showing the relationship between purpose of a loan and repaying the loan on time
loan_purpose_table = scoring_credit.pivot_table(index='loan_use', columns='defaulted_before',
                                                      values='debt', aggfunc='count', margins=True)

In [87]:
print('Pivot table showing relationship between purpose of a loan and repaying the loan on time:\n')

# make a column for subcategory representation in all entries
loan_purpose_table['% in tot entries'] = (loan_purpose_table['All'] / len(scoring_credit) * 100).round(1)

# if you get a loan for this purpose you have a _% chance of defaulting
loan_purpose_table['% chance of default'] = (loan_purpose_table['Yes'] / loan_purpose_table['All'] * 100).round(1)
# if you get a loan for this purpose you are _ times more likely to default
# loan_purpose_table['times more likely to default'] = (loan_purpose_table['No'] / loan_purpose_table['Yes']).round(1)

print(loan_purpose_table.head(10))

Pivot table showing relationship between purpose of a loan and repaying the loan on time:

defaulted_before     No   Yes    All  % in tot entries  % chance of default
loan_use                                                                   
car                3903   403   4306              20.1                  9.4
education          3643   370   4013              18.7                  9.2
house              3553   256   3809              17.8                  6.7
property           2348   190   2538              11.8                  7.5
real estate        4128   336   4464              20.8                  7.5
wedding            2138   186   2324              10.8                  8.0
All               19713  1741  21454             100.0                  8.1


### Conclusion

Most clients take out loans for buying a car or real estate. However if loans for buying houses, real estate, and property related expenses were combined, this would make up the majority.

Clients buying a house have the lowest chance of defaulting (6.7%).

Clients buying a car or paying for education have the highest chance of defaulting (9.4 & 9.2%).

Clients buying real estate or paying for property related expenses have the same chance of defaulting (7.5%).

Clients paying for a wedding have a higher chance of defaulting (8.0%) than those buying real estate/property (7.5%) and lower than those buying a car/paying for education (9.4 & 9.2%).

All reasons for the loan appears to have roughly the same chance of defaulting. This implies that the reason for the loan is not indicative of whether or not they will default, rather it may emphasize that no matter the reason the chance of default is the same.

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: I couldn't agree more!
</div>

## General conclusion

The investigation wanted to evaluate if the number of children and marital status impact the chance of defaulting on a loan, while also evaluating income level and purpose for the loan.

***Possible indicators of defaulting***
- The number of children a client has appears to be a good indicator of the chance of defaulting because the more children a client has the higher their chance of defaulting
- Income level also appears to be a good indicator because as income increases from 0 to <125,000 the chance of default goes down, 125,000 or more and it goes up.

***Unlikely indicators of defaulting***
- Marital status does not appear to be a good indicator because there is no obvious trend based on the relationship status of a client. It might be assumed that those with civil partnerships and married status would have a similar chance because there are potentially two contributing incomes, while those divorced, unmarried, and widowed would have a similar chance because they have potentially one contributing income. However this is not the case according to this data set.
- The purpose of the loan also did not have a clear trend. The results varied slightly but they all have about the same chance of defaulting.

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
Success: You are great at writing conclusions! It is really perfect! 
</div>

## Project Readiness Checklist

Put 'x' in the completed points. Then press Shift + Enter.

- [x]  file open;
- [x]  file examined;
- [x]  missing values defined;
- [x]  missing values are filled;
- [x]  an explanation of which missing value types were detected;
- [x]  explanation for the possible causes of missing values;
- [x]  an explanation of how the blanks are filled;
- [x]  replaced the real data type with an integer;
- [x]  an explanation of which method is used to change the data type and why;
- [x]  duplicates deleted;
- [x]  an explanation of which method is used to find and remove duplicates;
- [x]  description of the possible reasons for the appearance of duplicates in the data;
- [x]  data is categorized;
- [x]  an explanation of the principle of data categorization;
- [x]  an answer to the question "Is there a relation between having kids and repaying a loan on time?";
- [x]  an answer to the question "Is there a relation between marital status and repaying a loan on time?";
- [x]  an answer to the question "Is there a relation between income level and repaying a loan on time?";
- [x]  an answer to the question "How do different loan purposes affect on-time repayment of the loan?"
- [x]  conclusions are present on each stage;
- [x]  a general conclusion is made.